app/graphql/types/query_types/category_type.rb

```diff
@@ -7,17 +7,141 @@ class QueryTypes::CategoryType < Types::BaseObject
     field :parent, QueryTypes::CategoryType, null: true
     field :children, [QueryTypes::CategoryType], null: true
 
-    field :event_records_count, Integer, null: true
-    field :event_records, [QueryTypes::EventRecordType], null: true
-    field :generic_items_count, Integer, null: true
-    field :generic_items, [QueryTypes::GenericItemType], null: true
+    field :event_records_count,
+          Integer,
+          null: true,
+          resolve: lambda { |category, _args, ctx|
+            location = ctx.irep_node.parent.arguments.location
+
+            event_records = if location
+                              category.event_records.by_location(location)
+                            else
+                              category.event_records
+                            end
+
+            event_records.visible.count
+          }
+    field :event_records,
+          [QueryTypes::EventRecordType],
+          null: true,
+          resolve: lambda { |category, _args, ctx|
+            location = ctx.irep_node.parent.arguments.location
+
+            if location
+              category.event_records.by_location(location)
+            else
+              category.event_records
+            end
+          }
+    field :generic_items_count,
+          Integer,
+          null: true,
+          resolve: lambda { |category, _args, ctx|
+            location = ctx.irep_node.parent.arguments.location
+
+            generic_items = if location
+                              category.generic_items.by_location(location)
+                            else
+                              category.generic_items
+                            end
+
+            generic_items.visible.count
+          }
+    field :generic_items,
+          [QueryTypes::GenericItemType],
+          null: true,
+          resolve: lambda { |category, _args, ctx|
+            location = ctx.irep_node.parent.arguments.location
+
+            if location
+              category.generic_items.by_location(location)
+            else
+              category.generic_items
+            end
+          }
```

Comment:
any ideas how to reduce redundancies in this file? it is nearly the same for all fields for each type that can have a location filter.

Reply:
https://stackoverflow.com/questions/1347570/how-do-i-create-a-reusable-block-proc-lambda-in-ruby

Reply:
refactoring in the making #315